2018/002-update-channels.md

```diff
@@ -0,0 +1,184 @@
+lose support.
+
+To make a concrete example:
+1. The current version of CaaSP is 3.0
+    1. `bash` security fix is sent to the 3.0 channel
+    1. `kubernetes-salt` bug fixes are sent to the 3.0 channel
+1. A new version of Kubernetes is released:
+    1. The 3.1 channel containing the new kubernetes version is created
+    1. A ssl security fix is pushed both to the 3.0 and to the 3.1 channel
+    1. `kubernetes-salt` bug fixes are sent to the 3.1 channel and eventually to the 3.0 one
+1. *X* months passed since the creation of the 3.1 channel
+    1. A kernel security fix is pushed **only** to the 3.1 channel
+
+### Proposed change
+
+The admin will have the ability of choosing when to switch to the next channel.
+
+In case of an airgapped system, a link to the release notes of the update channel
+is displayed. A system with access to the web can fetch the release notes and
+display them in the UI directly. In both scenarios, an admin can evaluate the
+upcoming changes beforehand.
+
+The UI of velum needs to be adapted to enable a switch to a new channel.
+
+## Detailed RFC
+
+We can separate the updates into different channels. There are 2 types of channels:
+
+* update channel: normal maintenance updates, which are safe and non-intrusive
+* new version channel: version updates which can contain intrusive updates
+
+This RFC proposes:
+
+* a front-end change in the velum UI to enable the admin to switch to a new version
+channel
+* a back-end change in salt to perform the channel switch on all cluster nodes
+
+### Proposed change (Detailed)
+
+#### Tools/Services
+
+We need a set of tools/services to indicate, whenever a maintenance update or a new version
+update is ready to be installed. They should run on each worker node to be more independent
+and decoupled from velum. If velum would be the initiator of the update check, it might lead
+to inconsistent and long execution runs in larger clusters.
+
+* "check for updates" service on each worker
+  - reports availability of maintenance updates to velum via grains
+  - once per day e.g.
+  - decoupled from velum, runs independently, just reports
+* "new version" checker service on each worker
+  - check for new versions (migration targets)
+    - query RMT/SMT/SCC API if there is a new compatible version available for this CaaSP version (SUSEConnect library)
+    - report back to velum via grains
+  - get the link to the release notes
+  - once per week or longer timeframe
+  - decoupled from velum, runs independently, just reports
+
+#### User Interface
+
+The user interface of velum needs to be adapted minimally.
+
+* [indirect] display of update content through [a link to the] release notes
+* display of update history via velum
+* button to trigger the maintenance updates (what we currently have)
+* button to switch to the next channel in velum
+* separate UI for updates and upgrades (new migration target)
+
```

Comment:
I think everything is clear so far: we will have a background process running on each node that will report back the availability of a migration update. This is going to be done via a grain. The grain will be handled by Velum, which will show a new button to trigger the cluster migration.

What is missing right now from the RFC is how the migration will actually take place. I guess the migration will be handled by some cli tool (`transactional-update` ?) that is available on each node. Is that correct (cc @thkukuk )?

If that proves to be right (we have a cli tool taking care of the migration), I have the following questions:

* What are going to be the responsibilities of this tool? For example, is it going to take care of swapping the current update channel with the new one or is that something up to us?
* How are we going to invoke it? AFAIK salt doesn't play along really well with long running jobs (cc @kiall, @ereslibre, @inercia). Should we just rely on some asynchronous process that takes care of the migration and notifies us about the node being able to be rebooted by setting yet another grain?